File: app/src/debug/java/de/mchllngr/quickopen/base/DebugApp.java
package de.mchllngr.quickopen.base;

import android.app.Application;

import com.pandulapeter.beagle.Beagle;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import de.mchllngr.quickopen.util.DebugDrawer;
import timber.log.Timber;

/**
 * Base-class used for debug initializations.
 */
public abstract class DebugApp extends Application {

    private final DebugDrawer debugDrawer = new DebugDrawer(this);

    @Override
    public void onCreate() {
        super.onCreate();

        initTimber();
        debugDrawer.init();
    }

    /**
     * Initialises {@link Timber} with debug configuration
     */
    private void initTimber() {
        Timber.plant(new Timber.DebugTree() {

            // Add the line number to the tag
            @Override
            protected String createStackElementTag(@NonNull StackTraceElement element) {
                return super.createStackElementTag(element) + '#' + element.getLineNumber();
            }

            @Override
            protected void log(int priority, String tag, @NotNull String message, Throwable t) {
                Beagle.INSTANCE.log(message, tag, null);
                super.log(priority, tag, message, t);
            }
        });
    }
}

File: app/src/debug/java/de/mchllngr/quickero/base/DebugApp.kt
package de.mchllngr.quickero.base

import android.app.Application
import com.pandulapeter.beagle.log.BeagleLogger
import de.mchllngr.quickero.util.DebugDrawer
import timber.log.Timber
import timber.log.Timber.DebugTree

/**
 * Base-class used for debug initializations.
 */
open class DebugApp : Application() {

    @Suppress("LeakingThis") private val debugDrawer = DebugDrawer(this)

    override fun onCreate() {
        super.onCreate()
        initTimber()
        debugDrawer.init()
    }

    /**
     * Initialises [Timber] with debug configuration
     */
    private fun initTimber() {
        Timber.plant(object : DebugTree() {

            // Add the line number to the tag
            override fun createStackElementTag(element: StackTraceElement): String? {
                return super.createStackElementTag(element) + '#' + element.lineNumber
            }

            override fun log(
                priority: Int,
                tag: String?,
                message: String,
                t: Throwable?
            ) {
                BeagleLogger.log("[$tag] $message", "Timber", t?.stackTraceToString())
                super.log(priority, tag, message, t)
            }
        })
    }
}
